### ISSUE TITLE
Can't generate compilation DB on M1 Mac

### ISSUE THREAD

**Issue by xjtian** [envoyproxy/envoy#22217]
*Description*:

`/tools/vscode/refresh_compdb.sh` doesn't succeed on an M1 MBP due to toolchain selection failure. This happens if I run the script directly on the host and inside the devcontainer, although for different reasons. Unfortunately I don't have an Intel Mac to attempt a repro on.

Toolchain debug output on the *devcontainer*:

```
[output truncated for brevity]

INFO: ToolchainResolution: Target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Selected execution platform @envoy_build_tools//toolchains:rbe_linux_clang_platform, type @bazel_tools//tools/cpp:toolchain_type -> toolchain @envoy_build_tools//toolchains/configs/linux/clang/cc:cc-compiler-k8
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_freebsd_386-impl; mismatching values: freebsd, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_freebsd_amd64-impl; mismatching values: freebsd
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_freebsd_arm-impl; mismatching values: freebsd, arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_freebsd_arm64-impl; mismatching values: freebsd, arm64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_illumos_amd64-impl; mismatching values: illumos
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_ios_amd64-impl; mismatching values: ios
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_ios_arm64-impl; mismatching values: ios, arm64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_js_wasm-impl; mismatching values: js, wasm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_386-impl; mismatching values: x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: execution platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_amd64-impl; mismatching values: arm64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_arm-impl; mismatching values: arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_arm64-impl; mismatching values: arm64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_mips-impl; mismatching values: mips
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_mips64-impl; mismatching values: mips64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_mips64le-impl; mismatching values: mips64le
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_mipsle-impl; mismatching values: mipsle
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_ppc64-impl; mismatching values: ppc64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_ppc64le-impl; mismatching values: ppc
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_riscv64-impl; mismatching values: riscv64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_linux_s390x-impl; mismatching values: s390x
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_nacl_386-impl; mismatching values: nacl, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_nacl_amd64p32-impl; mismatching values: nacl, amd64p32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_nacl_arm-impl; mismatching values: nacl, arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_netbsd_386-impl; mismatching values: netbsd, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_netbsd_amd64-impl; mismatching values: netbsd
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_netbsd_arm-impl; mismatching values: netbsd, arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_netbsd_arm64-impl; mismatching values: netbsd, arm64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_openbsd_386-impl; mismatching values: openbsd, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_openbsd_amd64-impl; mismatching values: openbsd
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_openbsd_arm-impl; mismatching values: openbsd, arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_openbsd_arm64-impl; mismatching values: openbsd, arm64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_plan9_386-impl; mismatching values: plan9, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_plan9_amd64-impl; mismatching values: plan9
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_plan9_arm-impl; mismatching values: plan9, arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_solaris_amd64-impl; mismatching values: solaris
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_windows_386-impl; mismatching values: windows, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_windows_amd64-impl; mismatching values: windows
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_windows_arm-impl; mismatching values: windows, arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: Rejected toolchain @go_sdk//:go_windows_arm64-impl; mismatching values: windows, arm64
INFO: ToolchainResolution:   Type @io_bazel_rules_go//go:toolchain: target platform @envoy_build_tools//toolchains:rbe_linux_clang_platform: No toolchains found.
INFO: Repository jq_linux_amd64 instantiated at:
  /workspaces/envoy/WORKSPACE:25:25: in <toplevel>
  /workspaces/envoy/bazel/dependency_imports.bzl:54:27: in envoy_dependency_imports
  /build/envoy-compdb/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/external/aspect_bazel_lib/lib/repositories.bzl:37:25: in register_jq_toolchains
Repository rule jq_platform_repo defined at:
  /build/envoy-compdb/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/external/aspect_bazel_lib/lib/private/jq_toolchain.bzl:183:35: in <toplevel>
ERROR: /build/envoy-compdb/_bazel_vscode/2d35de14639eaad1ac7060a4dd7e3351/external/com_envoyproxy_protoc_gen_validate/BUILD:11:10: While resolving toolchains for target @com_envoyproxy_protoc_gen_validate//:protoc-gen-validate: no matching toolchains found for types @io_bazel_rules_go//go:toolchain
ERROR: Analysis of target '//tools/type_whisperer:types_cc_proto_validate' failed; build aborted: 
```

Toolchain debug output on the *local host*:

```
[output truncated for brevity]

INFO: ToolchainResolution:   Type @aspect_bazel_lib//lib:jq_toolchain_type: target platform @local_config_platform//:host: No toolchains found.
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @local_config_platform//:host: Rejected toolchain @go_sdk//:go_linux_s390x-impl; mismatching values: linux, s390x
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @local_config_platform//:host: Rejected toolchain @go_sdk//:go_nacl_386-impl; mismatching values: nacl, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @local_config_platform//:host: Rejected toolchain @go_sdk//:go_nacl_amd64p32-impl; mismatching values: nacl, amd64p32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @local_config_platform//:host: Rejected toolchain @go_sdk//:go_nacl_arm-impl; mismatching values: nacl, arm
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @local_config_platform//:host: Rejected toolchain @go_sdk//:go_netbsd_386-impl; mismatching values: netbsd, x86_32
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @local_config_platform//:host: Rejected toolchain @go_sdk//:go_netbsd_amd64-impl; mismatching values: netbsd, x86_64
INFO: ToolchainResolution:     Type @io_bazel_rules_go//go:toolchain: target platform @local_config_platform//:host: Rejected toolchain @go_sdk//:go_netbsd_arm-impl; mismatching values: netbsd, arm
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: Rejected toolchain @rules_rust//rust/private/dummy_cc_toolchain:dummy_cc_wasm32; mismatching values: wasm32
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: Rejected toolchain @rules_rust//rust/private/dummy_cc_toolchain:dummy_cc_wasm32; mismatching values: wasm32
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: Rejected toolchain @rules_rust//rust/private/dummy_cc_toolchain:dummy_cc_wasm32; mismatching values: wasm32
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: Rejected toolchain @rules_rust//rust/private/dummy_cc_toolchain:dummy_cc_wasm32; mismatching values: wasm32
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: Rejected toolchain @local_config_cc//:cc-compiler-armeabi-v7a; mismatching values: arm
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: Rejected toolchain @local_config_cc//:cc-compiler-armeabi-v7a; mismatching values: arm

[...]

ERROR: /Users/jtian/src/envoy/tools/code/BUILD:10:3: While resolving toolchains for target //tools/code:extensions_build_config: no matching toolchains found for types @aspect_bazel_lib//lib:jq_toolchain_type
ERROR: Analysis of target '//tools/code:extensions_build_config' failed; build aborted:
```

*Repro steps*:
- Open the project in VSCode with the Remote-Container extension activated, and run the "Refresh Compilation Database" task.
- Or, set up the local dev env on an m1 Mac and run `./tools/vscode/refresh_compdb.sh`
- I'm running this on an M1 Max MBP, MacOS 12.4

**Comment by xjtian** [comment]
I found https://github.com/aspect-build/bazel-lib/issues/174 for the local dev issue. Let me see what happens if I use `yq` instead of `jq` rules locally.

**Comment by phlax** [comment]
i think my hesitation - apart from just finding time to test it is that jq is (i believe) faster - which matters for large files

that said - if we can workaround the arm issue just now either by just switching it out, or by using arch selects i think lets do it

**Comment by keith** [comment]
Looks like we just need to update bazel-lib to include this commit? https://github.com/aspect-build/bazel-lib/commit/32882befdf20f3011482fb5971c5747f69e732ca

**Comment by keith** [comment]
https://github.com/envoyproxy/envoy/pull/22239

**Comment by ramaraochavali** [comment]
@keith Thanks for the fix. After your fix, I am getting the following error

```
      Compiling with an SDK that doesn't seem to exist: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk
      Please check your Xcode installation
      clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -fPIC -I/var/folders/c8/z52g4gc5679gygt8z0dx0jqm0000gn/T/tmpz82uci07/tools/deps/include -I/var/folders/c8/z52g4gc5679gygt8z0dx0jqm0000gn/T/tmpz82uci07/tools/deps/include/ncursesw -I/var/folders/c8/z52g4gc5679gygt8z0dx0jqm0000gn/T/tmpz82uci07/tools/deps/include/uuid -Werror=unguarded-availability-new -g0 -I/private/tmp/envoy-compdb/_bazel_rama.rao/076790d0a66b7f66620e924600fa5192/external/python3_10_aarch64-apple-darwin/include/python3.10 -Ibuild/temp.macosx-11.0-arm64-cpython-310/include -c build/temp.macosx-11.0-arm64-cpython-310/_sodium.c -o build/temp.macosx-11.0-arm64-cpython-310/build/temp.macosx-11.0-arm64-cpython-310/_sodium.o
      clang: warning: no such sysroot directory: '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk' [-Wmissing-sysroot]
      In file included from build/temp.macosx-11.0-arm64-cpython-310/_sodium.c:57:
      /private/tmp/envoy-compdb/_bazel_rama.rao/076790d0a66b7f66620e924600fa5192/external/python3_10_aarch64-apple-darwin/include/python3.10/Python.h:25:10: fatal error: 'stdio.h' file not found
      #include <stdio.h>
               ^~~~~~~~~
      1 error generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pynacl
ERROR: Failed to build one or more wheels
Traceback (most recent call last):
  File "/private/tmp/envoy-compdb/_bazel_rama.rao/076790d0a66b7f66620e924600fa5192/external/python3_10_aarch64-apple-darwin/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/private/tmp/envoy-compdb/_bazel_rama.rao/076790d0a66b7f66620e924600fa5192/external/python3_10_aarch64-apple-darwin/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/private/tmp/envoy-compdb/_bazel_rama.rao/076790d0a66b7f66620e924600fa5192/external/rules_python/python/pip_install/parse_requirements_to_bzl/extract_single_wheel/__main__.py", line 4, in <module>
    main()
  File "/private/tmp/envoy-compdb/_bazel_rama.rao/076790d0a66b7f66620e924600fa5192/external/rules_python/python/pip_install/parse_requirements_to_bzl/extract_single_wheel/__init__.py", line 57, in main
```

It seems it is trying to use /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk but I have /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk. 

Did you run in to this issue as well?

**Comment by keith** [comment]
Interesting I do see the same

**Comment by keith** [comment]
Ok here's another required fix https://github.com/envoyproxy/envoy/pull/22248

but even with this I find that liburing fails to build for me, as a temporary workaround I found commenting out this line https://github.com/envoyproxy/envoy/blob/1cf5603dc5239c92e5bc38ef321f59ccf6eabc6e/source/common/io/BUILD#L29 breaks the connection in the dependency graph and the file generates

**Comment by phlax** [comment]
@keith can we add a select on the `uring` dep so it is not included or perhaps just hack the `gen_compilation_database.py::is_compile_target` function

**Comment by keith** [comment]
We are actually doing a select on the target upstream that pulls in the library I linked above. So it's just a side effect of doing `source/...` that still pulls in this library. I'm actually surprised this is the only case of this today.

One option is to use this https://github.com/grailbio/bazel-compilation-database/pull/100 and mark the target as compatible only with Linux. I'm not sure what else would fall out of that today though, as we're not currently using `target_compatible_with` much

**Comment by keith** [comment]
And here's an example of what the incompatible solution would look like if we wanted to go that route https://github.com/envoyproxy/envoy/pull/22273, please ptal

**Comment by phlax** [comment]
trying to understand the implications of disabling it on all platforms - i guess no code highligting etc for that code

iiuc im also wondering if we could generate in a platform-independent way - it seems strange/arbitrary that someone editing on one arch would see the source code differently to another - irrespective of what was ultimate compiled/deployed

that said - i think it would be good to fix this asap (and maybe backport the fix) - so if we need to workaround in some way - probably, lets do it

**Comment by keith** [comment]
Yea I'm not sure if it could cascade to something more than just no completion for those 2 files or not. I think that's still the safest one for now https://github.com/envoyproxy/envoy/pull/22272

I think the ideal longer term solution is using https://github.com/envoyproxy/envoy/pull/22273 once the patch in bazel lands, in which case we would just annotate targets as needed and not have to do anything special for compdb specifically

**Comment by ramaraochavali** [comment]
@keith Thank you so much. With this now compilation database generates correctly. But in VS code it still says 
errors like `cstdint file not found` and `make_unique is not a member of 'std'`  - may be some setup issue. I have all the extensions that .devcontainer.json recommends. Any idea?

**Comment by keith** [comment]
Can you check the compile_commands.json file for the file you're working in to see if the `-std=` arg looks right??

**Comment by ramaraochavali** [comment]
-std=c++17 - This is what it has for all files.

**Comment by keith** [comment]
hrms seems right, you can try running one of those commands to see if you get that outside the editor as well

**Comment by phlax** [comment]
@keith im wondering if this is resolved and im also wondering whether we should backport this to the release/1.23 branch - not sure how likely someone is to be developing on that branch or how necessary this is to do so

**Comment by keith** [comment]
I think the core issue is fixed

**Comment by phlax** [comment]
cool - re backporting - im thinking lets skip it for now and if anyone complains we can backport it

**Comment by keith** [comment]
@ramaraochavali please file an new issue if you cant get to the bottom of that one!
